### PR TITLE
Added to the initial_nodes list all retrieved IP entries

### PR DESCRIPTION
**Pull request by barshaul**
Changed the cluster initialization to retrieve all IP entries from the initial nodes and use all resolved IPs.

### CODE REVIEW

File: redis/src/cluster_async/mod.rs [hunk 1]
```diff
@@ -675,15 +714,37 @@ where
...
+            let conn = match connections.get(addr).cloned() {
+                Some(conn) => Some(conn),
+                None => {
+                    // If it's a DNS endpoint, it could have been stored in the existing connections vector using the resolved IP address instead of the DNS endpoint's name.
+                    // We shall check if a connection is already exists under the resolved IP name.
+                    let (host, port) = match get_host_and_port_from_addr(addr) {
+                        Some((host, port)) => (host, port),
+                        None => return (addr, None),
+                    };
+                    match get_socket_addrs(host, port).await {
```
Comment (nihohit) [inline redis/src/cluster_async/mod.rs:733]: maybe `ok` and `find` can simplify the code here
```
get_socket_addrs(host, port).await.ok()
  .map(|socket_addresses|socket_addresses.find(|addr|connections.get(&addr.to_string()).cloned())
```

**Review by nihohit — APPROVED** [review]

File: redis/src/cluster_async/mod.rs [hunk 2]
```diff
@@ -1301,6 +1363,22 @@ where
...
+/// Splits a string address into host and port. If the passed address cannot be parsed, None is returned.
+/// [addr] should be in the following format: "<host>:<port>".
+fn get_host_and_port_from_addr(addr: &str) -> Option<(&str, u16)> {
+    let parts: Vec<&str> = addr.split(':').collect();
+    if parts.len() != 2 {
+        return None;
+    }
+    let host = parts.first().unwrap();
+    let port = parts.get(1).unwrap();
+    let port = match port.parse::<u16>() {
```
Comment (nihohit) [inline redis/src/cluster_async/mod.rs:1375]: maybe `ok` can simplify this
```rs
port.parse::<u16>().ok().map(|port|(host, port))
```

File: redis/src/cluster_async/mod.rs [hunk 3]
```diff
@@ -478,15 +478,54 @@ where
         Ok(connection)
     }
 
+    /// Go through each of the initial nodes and attempt to retrieve all IP entries from them.
+    /// If there's a DNS endpoint that directs to several IP addresses, add all addresses to the initial nodes list.
+    pub(crate) async fn try_to_expand_initial_nodes(
+        initial_nodes: &[ConnectionInfo],
+    ) -> Vec<String> {
+        stream::iter(initial_nodes)
```
[the remark this answers is not in the document]
Comment (barshaul) [reply redis/src/cluster_async/mod.rs:486]: NO

File: redis/src/cluster_async/mod.rs [hunk 4]
```diff
@@ -478,15 +478,54 @@ where
...
+            .fold(
```
[the remark this answers is not in the document]
Comment (barshaul) [reply redis/src/cluster_async/mod.rs:487]: We tried `flat_map`, and decided it's more easy to implement with `fold`

File: redis/src/cluster_async/mod.rs (hunk 1 above)
Comment (barshaul) [reply redis/src/cluster_async/mod.rs:733]: It would return the socket address, rather than the connection which I need

**Pull request merged**